constraint/pkg/apis/templates/transform.go

```diff
@@ -0,0 +1,69 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+)
+
+// AdAAddPreserveUnknownFields recurses through an *apiextensionsv1beta1.JSONSchemaProps
+// data structure, adding `x-kubernetes-preserve-unknown-fields: true` at every level
+// that type is equal to "object", "array", or is undefined.
+func AddPreserveUnknownFields(sch *apiextensionsv1beta1.JSONSchemaProps) error {
+	// Both an object and an array can have values not described in the schema.
+	// A blank Type could be anything, including an object or array, so we should
+	// treat it the same as those types.
+	if sch.Type == "" || sch.Type == "object" || sch.Type == "array" {
+		trueBool := true
+		sch.XPreserveUnknownFields = &trueBool
+	}
+
+	if sch.Properties != nil {
+		for k, v := range sch.Properties {
+			if err := AddPreserveUnknownFields(&v); err != nil {
+				return err
+			}
+
+			// As v is not a pointer, we need to set the contents of v back into the original data structure
+			sch.Properties[k] = v
+		}
+	}
+
+	if sch.Items != nil {
+		if sch.Items.Schema != nil {
+			if err := AddPreserveUnknownFields(sch.Items.Schema); err != nil {
+				return err
+			}
+		}
+
+		for k, v := range sch.Items.JSONSchemas {
+			if err := AddPreserveUnknownFields(&v); err != nil {
+				return err
+			}
+
+			// As v is not a pointer, we need to set the contents of v back into the original data structure
+			sch.Items.JSONSchemas[k] = v
+		}
+	}
+
+	if sch.AdditionalProperties != nil && sch.AdditionalProperties.Schema != nil {
```

Comment:
technically this should not be set, per the spec:

https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#specifying-a-structural-schema

So if it exists, this is already broken. Fortunately, I think they never allowed AdditionalProperties to be set? (hopefully...)

Reply:
AFAICT `additionalProperties` has the following restrictions on it (from that post):

> 3. does not set... additionalProperties... within an allOf, anyOf, oneOf or not, with the exception of the two pattern for x-kubernetes-int-or-string: true (see below).

And from lower in the article:

> The field additionalProperties cannot be set to false.
> The field additionalProperties is mutually exclusive with properties.

That makes it seem like `additionalProperties` _can_ be set, but not inside of those other validating directives like `allOf` and not alongside a `properties`.

Am I interpreting that correctly?

Reply:
Ah, you are, I missed that.